# 1173200618 赵沛霖 机器学习与人工智能 作业二

基于存在交互项多元线性回归模型预测学生被录取的概率

一、分析查看数据

In [1]:
import pandas as pd

df = pd.read_csv('Admission_Predict_Ver1.1.csv',index_col=0)
df.shape

(500, 8)

In [2]:
df.head(5)

GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  \
Serial No.                                                               
1                 337          118                  4  4.5   4.5  9.65   
2                 324          107                  4  4.0   4.5  8.87   
3                 316          104                  3  3.0   3.5  8.00   
4                 322          110                  3  3.5   2.5  8.67   
5                 314          103                  2  2.0   3.0  8.21   

            Research  Chance of Admit   
Serial No.                              
1                  1              0.92  
2                  1              0.76  
3                  1              0.72  
4                  1              0.80  
5                  0              0.65

In [3]:
X = df.iloc[:,:7]
X.sample(5)

GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  \
Serial No.                                                               
338               332          118                  5  5.0   5.0  9.47   
393               326          112                  4  4.0   3.5  9.12   
71                332          118                  5  5.0   5.0  9.64   
440               312          105                  2  1.5   3.0  8.46   
103               314          106                  2  4.0   3.5  8.25   

            Research  
Serial No.            
338                1  
393                1  
71                 1  
440                0  
103                0

In [4]:
y=df.iloc[:,-1]
y.sample(5)

Serial No.
63     0.54
347    0.47
391    0.64
468    0.78
474    0.67
Name: Chance of Admit , dtype: float64

二、进行线性回归预测

In [5]:
from sklearn.model_selection import train_test_split

X_tr, X_ts, y_tr, y_ts = train_test_split(X, y)
X_tr.shape, X_ts.shape

((375, 7), (125, 7))

In [6]:
import statsmodels.formula.api as smf
model = smf.OLS(y, X)
model_result = model.fit()

In [7]:
model_result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:       Chance of Admit    R-squared:                       0.991
Model:                            OLS   Adj. R-squared:                  0.991
Method:                 Least Squares   F-statistic:                     8178.
Date:                Tue, 23 Apr 2019   Prob (F-statistic):               0.00
Time:                        19:13:59   Log-Likelihood:                 635.01
No. Observations:                 500   AIC:                            -1256.
Df Residuals:                     493   BIC:                            -1227.
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
GRE Score            -0.0030      0.000     -8.613      0.000      -0.004      -0.002
TOEFL Score           0.0033      0.001      3.342      0.001       0.001       0.005
University Rating     0.0166      0.004      3.935      0.000       0.008       0.025
SOP                   0.0082      0.005      1.584      0.114      -0.002       0.018
LOR                   0.0171      0.005      3.622      0.000       0.008       0.026
CGPA                  0.1340      0.011     12.205      0.000       0.112       0.156
Research              0.0561      0.007      8.099      0.000       0.042       0.070
==============================================================================
Omnibus:                       79.089   Durbin-Watson:                   0.881
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              128.537
Skew:                          -0.979   Prob(JB):                     1.23e-28
Kurtosis:                       4.529   Cond. No.                     1.22e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.22e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

分析：模型拟合程度非常好Prob (F-statistic)接近于0。在这几个自变量中，SOP的P值大于0.05，因此需要把该项去掉，同时可能存在交互项，需要排除交互项的相互干扰。

In [8]:
tmp=df['University Rating'].copy()
df['UR']=tmp

del tmp

In [9]:
tmp=df['TOEFL Score'].copy()
df['TS']=tmp

del tmp

In [10]:
tmp=df['GRE Score'].copy()
df['GS']=tmp

del tmp

In [11]:
tmp=df['LOR '].copy()
df['LOR']=tmp

del tmp

In [12]:
tmp=df['Chance of Admit '].copy()
df['COA']=tmp

del tmp

In [13]:
df.head(5)

GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  \
Serial No.                                                               
1                 337          118                  4  4.5   4.5  9.65   
2                 324          107                  4  4.0   4.5  8.87   
3                 316          104                  3  3.0   3.5  8.00   
4                 322          110                  3  3.5   2.5  8.67   
5                 314          103                  2  2.0   3.0  8.21   

            Research  Chance of Admit   UR   TS   GS  LOR   COA  
Serial No.                                                       
1                  1              0.92   4  118  337  4.5  0.92  
2                  1              0.76   4  107  324  4.5  0.76  
3                  1              0.72   3  104  316  3.5  0.72  
4                  1              0.80   3  110  322  2.5  0.80  
5                  0              0.65   2  103  314  3.0  0.65

In [14]:
import statsmodels.formula.api as smf
model = smf.ols('COA~GS+TS+LOR+CGPA+Research+UR+SOP+UR:SOP', df).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    COA   R-squared:                       0.825
Model:                            OLS   Adj. R-squared:                  0.823
Method:                 Least Squares   F-statistic:                     290.1
Date:                Tue, 23 Apr 2019   Prob (F-statistic):          1.24e-180
Time:                        19:14:05   Log-Likelihood:                 706.30
No. Observations:                 500   AIC:                            -1395.
Df Residuals:                     491   BIC:                            -1357.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.1781      0.108    -10.908      0.000      -1.390      -0.966
GS             0.0018      0.000      3.586      0.000       0.001       0.003
TS             0.0028      0.001      3.215      0.001       0.001       0.004
LOR            0.0163      0.004      3.975      0.000       0.008       0.024
CGPA           0.1173      0.010     12.189      0.000       0.098       0.136
Research       0.0245      0.007      3.748      0.000       0.012       0.037
UR            -0.0192      0.009     -2.159      0.031      -0.037      -0.002
SOP           -0.0188      0.008     -2.369      0.018      -0.034      -0.003
UR:SOP         0.0073      0.002      3.124      0.002       0.003       0.012
==============================================================================
Omnibus:                      103.125   Durbin-Watson:                   0.795
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              228.585
Skew:                          -1.083   Prob(JB):                     2.31e-50
Kurtosis:                       5.506   Cond. No.                     1.36e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.36e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

分析：University Rating和Statement of Purpose and Letter of Recommendation Strength存在交互影响，在排除影响之后两者P值均小于0.05，通过检验。

三、进行预测

In [15]:
y_hat = model_result.predict(X)
y_hat = pd.DataFrame(y_hat, columns=['pred'])
df_merge = df.merge(y_hat, left_index=True, right_index=True)
df_merge.head()

GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  \
Serial No.                                                               
1                 337          118                  4  4.5   4.5  9.65   
2                 324          107                  4  4.0   4.5  8.87   
3                 316          104                  3  3.0   3.5  8.00   
4                 322          110                  3  3.5   2.5  8.67   
5                 314          103                  2  2.0   3.0  8.21   

            Research  Chance of Admit   UR   TS   GS  LOR   COA      pred  
Serial No.                                                                 
1                  1              0.92   4  118  337  4.5  0.92  0.906497  
2                  1              0.76   4  107  324  4.5  0.76  0.800509  
3                  1              0.72   3  104  316  3.5  0.72  0.656187  
4                  1              0.80   3  110  322  2.5  0.80  0.734807  
5                  0              0.65   2  103  314  3.0  0.65  0.597560

In [16]:
import numpy as np
y = df['COA']
y_hat=df_merge['pred']
model_RMSE = np.sqrt(np.mean(np.square(y_hat-y)))
model_RMSE

0.06795205596095126

总结：此次作业运用存在交互项的多元线性回归模型进行预测。均方误差MSE很小，同时模型的R值为0.825，说明模型精度很高。模型的P值非常小，通过假设检验，模型通过检验；同时，每一个自变量的P值都非常小，都通过检验，而且排除了University Rating和SOP的相互影响。